include/envoy/singleton/manager.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+#include "envoy/singleton/instance.h"
+
+namespace Envoy {
+namespace Singleton {
+
+/**
+ * An abstract registration for a singleton entry.
+ */
+class Registration {
+public:
+  virtual ~Registration() {}
+  virtual std::string name() PURE;
+};
+
+/**
+ * A concrete implementation of a singleton registration. All singletons are referenced by name
+ * and must be statically registered ahead of time. This can be done like so:
+ *
+ * static constexpr char foo_singleton_name[] = "foo_singleton";
+ * static Registry::RegisterFactory<Singleton::RegistrationImpl<foo_singleton_name>,
```

Comment:
Can you comment on the tradeoffs of using a string in the template vs. a concrete type as done in `Registry::RegisterFactory`?